cmd/controller/wire.go

```diff
@@ -0,0 +1,52 @@
+// +build wireinject
+
+/*
+Copyright 2020 Google LLC.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+
+	"github.com/google/knative-gcp/pkg/apis/configs/gcpauth"
+	"github.com/google/knative-gcp/pkg/reconciler/events/auditlogs"
+	"github.com/google/knative-gcp/pkg/reconciler/events/build"
+	"github.com/google/knative-gcp/pkg/reconciler/events/pubsub"
+	"github.com/google/knative-gcp/pkg/reconciler/events/scheduler"
+	"github.com/google/knative-gcp/pkg/reconciler/events/storage"
+	"github.com/google/knative-gcp/pkg/reconciler/identity/iam"
+	"github.com/google/knative-gcp/pkg/reconciler/intevents/pullsubscription/keda"
+	"github.com/google/knative-gcp/pkg/reconciler/intevents/pullsubscription/static"
+	"github.com/google/knative-gcp/pkg/reconciler/intevents/topic"
+	"github.com/google/knative-gcp/pkg/reconciler/messaging/channel"
+	"github.com/google/wire"
+	"knative.dev/pkg/injection"
+)
+
+func InitializeControllers(ctx context.Context) ([]injection.ControllerConstructor, error) {
+	panic(wire.Build(
+		Controllers,
+		ClientOptions,
+		iam.PolicyManagerSet,
+		wire.Struct(new(gcpauth.StoreSingleton)),
```

Comment:
Is this the right way to do a singleton? What I want is to create a single gcpauth.Store. Because I only seem to be able to do that while the various controllers are being created (I am using the passed in configmap.Watcher, I suppose I could just make my own here instead), I made a wrapping singleton struct that uses sync.Once to make sure we only make one of them.

Reply:
I think this is a good way to do this. Of course it would be ideal if the cmw could be injected by wire as well but that will require significant changes to controller injection.